src/sql/base/browser/ui/table/table.ts

```diff
@@ -121,6 +123,58 @@ export class Table<T extends Slick.SlickData> extends Widget implements IDisposa
 		this.mapMouseEvent(this._grid.onHeaderClick, this._onHeaderClick);
 		this.mapMouseEvent(this._grid.onDblClick, this._onDoubleClick);
 		this._grid.onColumnsResized.subscribe(() => this._onColumnResize.fire());
+
+		this._grid.onClick.subscribe((e, data) => {
+			if (this.isTreeGrid()) {
+				this.toggleTreeGridParent(data.row, data.cell);
+			}
+		});
+
+		this._grid.onKeyDown.subscribe((e, data) => {
+			if (this.isTreeGrid()) {
+				if ((<any>e).keyCode === 13) {
+					// When we press enter on an collapsed cell we expand it.
+					this.toggleTreeGridParent(data.row, data.cell);
+				} else if ((<any>e).keyCode === 37) {
+					// Left arrow on first cell of the expanded row collapses it
+					if (data.cell === 0) {
+						this.toggleTreeGridParent(data.row, this.expandableColumnIndex(), false);
+					}
+				} else if ((<any>e).keyCode === 39) {
+					// Right arrow on last cell of the collapsed row expands it.
+					if (data.cell === (this._grid.getColumns().length - 1)) {
+						this.toggleTreeGridParent(data.row, this.expandableColumnIndex(), true);
+					}
+				}
+			}
+		});
+
+		this._grid.onRendered.subscribe((e, data) => {
+			if (this.isTreeGrid()) {
+				// Changing table role from grid to treegrid
+				this._tableContainer.setAttribute('role', 'treegrid');
+				for (let i = 0; i < this._data.getLength(); i++) {
```

Comment:
this is not efficient; you should only process the visible rows:  use this.grid.getViewPort() method to get the view port, use the top and bottom properties as start/end rows.